sdks/java/core/src/main/java/org/apache/beam/sdk/util/RowStringInterpolator.java

```diff
@@ -0,0 +1,192 @@
+ *
+ * <p>Nested fields can be specified using dot-notation (e.g. {@code "top.middle.nested"}).
+ *
+ * <p>Configure a {@link RowStringInterpolator} like so:
+ *
+ * <pre>{@code
+ * String template = "unified {foo} and {bar.baz}!";
+ * Row inputRow = {foo: "batch", bar: {baz: "streaming"}, ...};
+ *
+ * RowStringInterpolator interpolator = new RowStringInterpolator(template, beamSchema);
+ * String output = interpolator.interpolate(inputRow);
+ * // output --> "unified batch and streaming!"
+ * }</pre>
+ *
+ * <p>Additionally, {@link #interpolate(Row, BoundedWindow, PaneInfo, Instant)} can be used in
+ * streaming scenarios to substitute windowing metadata into the template String. To make use of
+ * this, use the relevant placeholder:
+ *
+ * <ul>
+ *   <li>$WINDOW: the window's string representation
+ *   <li>$PANE_INDEX: the pane's index
+ *   <li>$YYYY: the element timestamp's year
+ *   <li>$MM: the element timestamp's month
+ *   <li>$DD: the element timestamp's day
+ * </ul>
+ *
+ * <p>For example, your Sting template can look like:
+ *
+ * <pre>{@code "unified {foo} and {bar} since {$YYYY}-{$MM}!"}</pre>
+ */
+public class RowStringInterpolator implements Serializable {
+  private final String template;
+  private final Set<String> fieldsToReplace;
+  // Represents the string representation of the element's window
+  public static final String WINDOW = "$WINDOW";
+  public static final String PANE_INDEX = "$PANE_INDEX";
+  // Represents the element's pane index
+  public static final String YYYY = "$YYYY";
+  public static final String MM = "$MM";
+  public static final String DD = "$DD";
+  private static final Set<String> WINDOWING_METADATA =
+      Sets.newHashSet(WINDOW, PANE_INDEX, YYYY, MM, DD);
+  private static final Pattern TEMPLATE_PATTERN = Pattern.compile("\\{(.+?)}");
+
+  /**
+   * @param template a String template, potentially with placeholders in the form of curly braces,
+   *     e.g. {@code "my {foo} template"}. During interpolation, these placeholders are replaced
+   *     with values in the Beam Row. For more details and examples, refer to the top-level
+   *     documentation.
+   * @param rowSchema {@link Row}s used for interpolation are expected to be compatible with this
+   *     {@link Schema}.
+   */
+  public RowStringInterpolator(String template, Schema rowSchema) {
+    this.template = template;
+
+    Matcher m = TEMPLATE_PATTERN.matcher(template);
+    fieldsToReplace = new HashSet<>();
+    while (m.find()) {
+      fieldsToReplace.add(StringUtils.strip(m.group(), "{}"));
+    }
+
+    List<String> rowFields =
+        fieldsToReplace.stream()
+            .filter(f -> !WINDOWING_METADATA.contains(f))
+            .collect(Collectors.toList());
+
+    RowFilter.validateSchemaContainsFields(rowSchema, rowFields, "string interpolation");
+  }
+
+  /** Performs string interpolation on the template using values from the input {@link Row}. */
+  public String interpolate(Row row) {
```

Comment:
What is the motivation to have a separate no-windowing implementation?

Reply:
Figured why not, but on a second thought will remove it to keep windowing properties as first-class citizens